content/HackTheVote_2024/The Thirty-Twodle Challenge.md

```diff
@@ -0,0 +1,178 @@
+machine and help us predict their each and every future story?
+
+## Intuition
+This write-up is going to be less of a technical one because I (MettleSphee) took a weird approach: I like thinking about how games would work technically by looking at how they behave. Therefore, I found the directions to the flag without looking myself at the code, but the help I've gotten was incredible. I couldn't have written such amazing scripts/solutions for the challenge, and the credit goes to (Zenbassi) for basically doing the harder part of the challenge :) .
+
+## Write-up
+My teammate was doing the whole reversing process, while I chimed in and just wanted to try and run the binary to see what happens.
+
+Alright, who knows Wordle? The game has the player try to guess a 5 letter word, while allowing a total of 6 guesses. While guessing, each letter is marked in the word as follows:
+- Gray = the letter doesn't exist in that word;
+- Yellow = the letter exists in the word, but it's not in the right position;
+- Green = the letter is correct and in the right position;
+
+Sounds simple, right? Well, here's where the twist comes in: You have to guess 32 words in one go. Now it feels like a challenge.
+But don't worry, it's not *that* hard. This game has some specific caveats:
+- The executable asks you for a seed; this means the words to guess are the same every run on a specific seed (and therefore, can be guessed after multiple rounds);
+- Whenever you have all the correct letters for (1) word, it will do something called auto-solve; This time, instead of the word being highlighted in green, we have a new color:
+  - Purple = the word has been auto-solved by the program;
+- Auto-solves *consume* 1 guess from the total number of guesses;
+- You have 37 guesses to find all the words;
+
+That's a lot of information. Okay, so that means we just have to input all words and we win, right?
+![[wordle1.png]]
+Guess not. By looking through the code, it gives access to ``/bin/sh`` only when (simplified) all words are auto-solved. Given that it adds a guess whenever it auto-solves a word, we basically have 37 - 32 = 5 guesses which we can use to add words that contribute to auto-solving. How do we do that?
+
+We can extract the full wordlist from the binary, and we can get the solution for that seed. As it turns out, finding the right words to auto-solve for a specific seed *may* be difficult. Factors which include having too many letters. As we have only 5 guesses with 5 letters each, that means a total of 25 letters. Given that we don't have many vowels, or other characters that repeat more often than not, we're going to have to find a good seed. The words are taken in a pseudo-random order from the wordlist using libc's ``srand(seed)`` function.
+
+(Zenbassi) wrote the scripts to do the following things (simplified):
+- Find a seed with a score less than 5; the lower score the easier result;
```

Comment:
Add smth like this as an explanation for that score: 

For each word in the solution we computed a dictionary with the letter frequency. We then computed the union of these dictionaries, taking the maximum from the common values. The score is the sum of the values in the union divided by 5. It represents something like _the average number of unique letters from the solution alphabet, that each of our chosen words is expected to contain_. This number is relevant, because a score greater than 5 will mean that we cannot find a solution, given that we're limited to words of size 5.

smth like that